ocaml/xapi/xapi_pool_helpers.ml

```diff
@@ -269,3 +269,308 @@ let apply_guest_agent_config ~__context =
         (Printexc.to_string e)
   in
   call_fn_on_slaves_then_master ~__context f
+
+module PeriodicUpdateSync = struct
+  let periodic_update_sync_task_name = "Periodic update synchronization"
+
+  let seconds_random_within_an_hour () =
+    let secs_of_an_hour = 60 * 60 in
+    let secs_random = Random.int secs_of_an_hour in
+    debug "[PeriodicUpdateSync] seconds_random_within_an_hour: %d" secs_random ;
+    secs_random
+
+  let weekday_int_to_str ~weekday_int =
+    match weekday_int with
+    | 0 ->
+        "Sunday"
+    | 1 ->
+        "Monday"
+    | 2 ->
+        "Tuesday"
+    | 3 ->
+        "Wednesday"
+    | 4 ->
+        "Thursday"
+    | 5 ->
+        "Friday"
+    | 6 ->
+        "Saturday"
+    | _ ->
+        "Wrong weekday number, it should be 0 to 6, where 0 is Sunday"
+
+  let frequency_to_str ~frequency =
+    match frequency with
+    | `daily ->
+        "daily"
+    | `monthly ->
+        "monthly"
+    | `weekly ->
+        "weekly"
+
+  let periodic_update_sync_delay ~__context ~first_run =
+    let frequency =
+      Db.Pool.get_update_sync_frequency ~__context
+        ~self:(Helpers.get_pool ~__context)
+    in
+    let day_configed =
+      Db.Pool.get_update_sync_day ~__context ~self:(Helpers.get_pool ~__context)
+    in
+    let hour_configed =
+      Db.Pool.get_update_sync_hour ~__context ~self:(Helpers.get_pool ~__context)
+    in
+    debug
+      "[PeriodicUpdateSync] periodic_update_sync_delay first_run=%B: \
+       frequency=%s, day_configed=%Ld, hour_configed=%Ld"
+      first_run
+      (frequency_to_str ~frequency)
+      day_configed hour_configed ;
+    let day_configed_int = Int64.to_int day_configed in
+    let hour_configed_int = Int64.to_int hour_configed in
+    let secs_now = Unix.time () in
+    let tm_now = Unix.gmtime secs_now in
+    let secs_random_within_an_hour = seconds_random_within_an_hour () in
+    match frequency with
```

Comment:
A common logic could be:
1. get the current time
2. get the current interval ([s1, e1) and schedule offset1  (day, week, month)
3. get the next interval ([s2, e2) and schedule offset2  (day, week, month)
4. if current time < e1, return e1 - current time
5. else, return (s2  + current time + offset2)

So for each kind of schedule (daily, weekly, monthly), the problem is only to calculate an interval for current or next.